### PR TITLE
[CBRD-24191] Refactoring function 'dwb_adjust_write_buffer_values'.

### PR DESCRIPTION
**Pull request by hyahong**
http://jira.cubrid.org/browse/CBRD-24191

Purpose
Refactor the code to be a little more intuitive and consistent.

Implementation
N/A

Remarks
N/A

### CODE REVIEW

File: src/storage/double_write_buffer.c
Comment (wrlawodms) [inline src/storage/double_write_buffer.c]: It looks correct now, but this seems to assume that `DWB_MIN_BLOCKS` is 1. If then, I have a comment. The value is defined separately, which means it should be decoupled with this code. We should be able to change the value without considering this code. And to use `min_size` to set this, it'd be better to `assert (IS_POWER_OF_2(min_size))`.
```suggestion
      assert (IS_POWER_OF_2(min_size));
      unsigned int block_limit = min_size << 1;
```

**Review by wrlawodms — COMMENTED** [review]
Looks better and correct. Good job. 👍
I have an opinion, though.

**Comment by hyahong** [comment]
@wrlawodms 
I couldn't think that the value should be decoupled.
Also, including the parts you suggested, it seems that assertion about min_size and max_size being a power of 2 is necessary.
Thank you!

**Review by wrlawodms — APPROVED** [review]
LGTM. Thank you for your continued contribution.
It seems to require a simple test case. @hornetmj will decide it.

File: src/storage/double_write_buffer.c
Comment (hornetmj) [inline src/storage/double_write_buffer.c]:
```suggestion
      while (dwb_limit < *p_double_write_buffer_size)
```

Comment (hornetmj) [reply src/storage/double_write_buffer.c]: "dwb_limit >= DWB_MAX_SIZE" is never happen.

File: src/storage/double_write_buffer.c
Comment (hornetmj) [inline src/storage/double_write_buffer.c]:
```suggestion
  assert (IS_POWER_OF_2 (*p_double_write_buffer_size) && *p_double_write_buffer_size >= min_size && *p_double_write_buffer_size <= max_size);
```

File: src/storage/double_write_buffer.c
Comment (hornetmj) [inline src/storage/double_write_buffer.c]: *p_num_blocks is always in the range min_size to max_size. (min_size < *p_num_blocks < max_size) So, "block_limit < DWB_MAX_BLOCKS &&" can be removed.

File: src/storage/double_write_buffer.c
Comment (hornetmj) [inline src/storage/double_write_buffer.c]: In my opinion, it is better to check '*p_num_blocks < min_size' too. It is more readable even if it is already checked in the debug mode by assert.

**Review by hornetmj — APPROVED** [review]

File: src/storage/double_write_buffer.c
Comment (joohok) [inline src/storage/double_write_buffer.c]: is this assertion required? 
min_size and max_size are initialized using constant values ​​that satisfy the 'assert' condition.

And I think this min/max_size variable refers to the size of the buffer, not the number of blocks. So the assert should check if it is a multiple of 512 * 1024.

File: src/storage/double_write_buffer.c
Comment (joohok) [inline src/storage/double_write_buffer.c]: How about replacing `min/max_size` variables with `min/max_block` ?
I think it is more readable, and also is this assertion required? 

if the check is required, I think it is required to be mentioned when defining `DWB_MAX/MIN_BLOCKS` , `DBW_MAX/MIN_SIZE` at the top of `double_writer_buffer.c`.

**Comment by hornetmj** [comment]
@Yechan0815 The behavior for determining *p_double_write_buffer_size and *p_num_blocks is exactly the same. Since it is the same pattern, How about creating a new function for it?

I recommend the following method. How about introducing the dwb_load_buffer_size() and dwb_load_block_count() functions instead of the dwb_adjust_write_buffer_values()? You can also remove the dwb_adjust_write_buffer_values(). Also, create a new function that takes the MIN/MAX value and returns it into a power of two.

# dwb_load_buffer_size(&double_write_buffer_size)
- double_write_buffer_size = prm_get_integer_value (PRM_ID_DWB_SIZE);
- check the double_write_buffer_size whether or not 0
- process other part as a function
# dwb_load_block_count(&num_blocks)
- num_blocks = prm_get_integer_value (PRM_ID_DWB_BLOCKS);
- check the num_blocks whether or not 0
- process other part as the function

**Comment by hyahong** [comment]
@hornetmj 
It is definitely a effective structure. I will try it!

**Comment by hyahong** [comment]
@joohok 
Because all calculations are done under the assumption that min/max_size is power of 2, I added debug assertion just in case. I will reflect your suggestion in structure to be changed!

**Comment by eido5** [comment]
> @Yechan0815 The behavior for determining *p_double_write_buffer_size and *p_num_blocks is exactly the same. Since it is the same pattern, How about creating a new function for it?
> 
> I recommend the following method. How about introducing the dwb_load_buffer_size() and dwb_load_block_count() functions instead of the dwb_adjust_write_buffer_values()? You can also remove the dwb_adjust_write_buffer_values(). Also, create a new function that takes the MIN/MAX value and returns it into a power of two.
> 
> # dwb_load_buffer_size(&double_write_buffer_size)
> * double_write_buffer_size = prm_get_integer_value (PRM_ID_DWB_SIZE);
> * check the double_write_buffer_size whether or not 0
> * process other part as a function
> 
> # dwb_load_block_count(&num_blocks)
> * num_blocks = prm_get_integer_value (PRM_ID_DWB_BLOCKS);
> * check the num_blocks whether or not 0
> * process other part as the function

I may guess we can make "DWB_adjust()" using by overload in C++.
And then we can use just 1 function. because it is mentioned as "it works exaclty same".
And we can shorten the code length.
I cannot implement it right now as pseudo code but try to make pseudo code.

**Comment by hyahong** [comment]
@hornetmj I changed it first. I don't know if it's okay.

**Comment by hyahong** [comment]
@eido5 
Is it this kind of structure? This is a code that has been slightly changed in the current commit, and it has been used just one function.

(I'm attaching an image because I don't know how to upload the code changed without changing the commit.)

**Comment by eido5** [comment]
> 

Totally nice.
In my opinion, it is more readable code and simple.
How about this implementation? @hornetmj , @wrlawodms , @joohok

File: src/storage/double_write_buffer.c
Comment (joohok) [inline src/storage/double_write_buffer.c:772]: If the reason you need assert() here is that `DWB_MIN_SIZE` and `DWB_MAX_SIZE`are defined outside of this code, then I think you also need to compare the sizes of `DWB_MIN_SIZE` and `DWB_MAX_SIZE` and check if `DWB_MIN_SIZE` is greater than 512K. what do you think?
- DWB_MAX_SIZE >= DWB_MIN_SIZE
- DWB_MIN_SIZE >= 512 * 1024

And how about mentioning these conditions where they are defined?

**Comment by hyahong** [comment]
@joohok 
Values are defined by people, so I added a debug assertion to make sure that the values are valid.
adding the above condition such as "DWB_MAX_SIZE >= DWB_MIN_SIZE" makes it seem clearer.
If the structure doesn't change anymore, I will add conditions to dwb_load_buffer_size and dwb_load_block_count!

and sorry, I couldn't understand exactly what "mentioning these conditions where they are defined" means.
Could you please explain a little more in detail?

File: src/storage/double_write_buffer.c
Comment (joohok) [reply src/storage/double_write_buffer.c:772]: @Yechan0815  
Please, reply to this comment. 

> @joohok
Values are defined by people, so I added a debug assertion to make sure that the values are valid.
adding the above condition such as "DWB_MAX_SIZE >= DWB_MIN_SIZE" makes it seem clearer.
If the structure doesn't change anymore, I will add conditions to dwb_load_buffer_size and dwb_load_block_count!

I understand what you mean, it will be nice! 

> and sorry, I couldn't understand exactly what "mentioning these conditions where they are defined" means.
Could you please explain a little more in detail?

I mean, it will be better if there is some comment where these values are defined. it is just an opinion.
```
#define DWB_MAX_SIZE          .... /* add some comment that this value is required to be power of 2 */
#define DWB_MIN_SIZE           .... /* add some comment that this value is required to be power of 2  and greater than 512K */
```

**Review by joohok — APPROVED** [review]
Nice work, Thanks for contribution

File: src/storage/double_write_buffer.c
Comment (wrlawodms) [inline src/storage/double_write_buffer.c]:
```suggestion
 * return   : true or false.
```

File: src/storage/double_write_buffer.c
Comment (wrlawodms) [inline src/storage/double_write_buffer.c]:
```suggestion
 * return   : true or false.
```

File: src/storage/double_write_buffer.c
Comment (wrlawodms) [inline src/storage/double_write_buffer.c:726]: I think that this comment should be alive on `dwb_load_buffer_size()`. The buffer size must be a multiple of 512k, not only a power of 2.

File: src/storage/double_write_buffer.c
Comment (wrlawodms) [inline src/storage/double_write_buffer.c]: I want it to be more meaningful so that it says the purpose or role of this function. For example, It can be `dwb_adjust_[something]_value()`, `dwb_power2_ceil()` or something. I suppose you can make a better choice than me.

**Review by wrlawodms — COMMENTED** [review]
review on the changed code.

File: src/storage/double_write_buffer.c
Comment (hornetmj) [inline src/storage/double_write_buffer.c]:
```suggestion
    }

  dwb_adjust_value (DWB_MIN_SIZE, DWB_MAX_SIZE, p_double_write_buffer_size);

  return true;
```

File: src/storage/double_write_buffer.c
Comment (hornetmj) [inline src/storage/double_write_buffer.c]:
```suggestion

  dwb_adjust_value (DWB_MIN_BLOCKS, DWB_MAX_BLOCKS, p_num_blocks);

  return true;
```

**Review by hornetmj — APPROVED** [review]

**Comment by hyahong** [comment]
@wrlawodms @joohok It's corrected. Thank you!

**Review by eido5 — APPROVED** [review]

**Comment by hornetmj** [comment]
@Yechan0815 First, sorry for the late reply.

This is a refactoring and we expect that there is no execution change after merging this PR. So, I wonder how do you process the developer test to verify this. Please let me know when you are ready. I will make test scenarios but I want to know it.

**Comment by hyahong** [comment]
@hornetmj 
First of all, I confirmed the same with the code flow.
And I verified it by debugger, setting values smaller than min, min value, values between min and max, max value, and values larger than max.
(Not all values between min and max are checked.)

+ I modified dwb, And I don't know why sql fails. Can I get some advice?

**Comment by hornetmj** [comment]
@Yechan0815 The fails of the test are not related to your changes. So, don't worry about that. And I will merge this PR when we prepared the test scenarios to verify this. Thanks to share how you do the test.

**Comment by hornetmj** [comment]
@mhoh3963 Please merge this PR. The failures of the CircleCI are not related to this PR.

**Comment by hornetmj** [comment]
@mhoh3963 Please merge this PR. The failures of the CircleCI are not related to this PR. I am re-requesting because it seems that the previous request has not been confirmed.

**Pull request merged**
